places/src/api/matcher.rs

```diff
@@ -0,0 +1,303 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use rusqlite::{
+    self,
+    types::{FromSql, FromSqlError, FromSqlResult, ToSql, ToSqlOutput, ValueRef},
+};
+use url::Url;
+
+use db::PlacesDb;
+use error::{ErrorKind, Result};
+
+const MAX_RESULTS: usize = 10;
+
+#[derive(Hash, Eq, PartialEq, Clone, Copy)]
+pub enum SearchOption {
+    EnableActions,
+    DisablePrivateActions,
+    PrivateWindow,
+    UserContextId(i64),
+}
+
+/// A matcher returns autocomplete matches for a query string. Given a query
+/// and options, a matcher tokenizes the query, passes the tokens to all its
+/// registered providers, finds matches, filters them using a set of criteria,
+/// and returns them. A provider can be anything that returns URL suggestions:
+/// Places history, bookmarks, and keywords, synced tabs, search engine
+/// suggestions, and extension keywords.
+pub struct Matcher<'conn> {
+    conn: &'conn PlacesDb,
+}
+
+impl<'conn> Matcher<'conn> {
+    /// Synchronously queries all providers for autocomplete matches, given a
+    /// query string and options. This isn't cancelable yet; once a search is
+    /// started, it can't be interrupted, even if the user moves on (see
+    /// https://github.com/mozilla/application-services/issues/265).
+    pub fn search<Q: AsRef<str>>(&self, query: Q, options: &[SearchOption]) -> Result<Vec<Match>> {
+        /// TODO: Tokenize the query.
+        let matches = Vec::new();
+
+        let adaptive = Adaptive::new(query.as_ref(), self.conn, MAX_RESULTS);
+        let adaptive_matches = adaptive.search()?;
+
+        let suggestions = Suggestions::new(query.as_ref(), self.conn, MAX_RESULTS);
+        let suggestions_matches = suggestions.search()?;
+
+        Ok(matches)
+    }
+}
+
+/// The match reason specifies why an autocomplete search result matched a
+/// query. This can be used to filter and sort matches.
+pub enum MatchReason {
+    QueryString,
+    Path,
+    Origin,
+    Label,
+    PreviousUse,
+    Boundary,
+    Fuzzy,
+    Tags(String),
+}
+
+pub enum MatchLabel {
+    BookmarkTag,
+    Tag,
+    Bookmark,
+    SwitchTab,
+    Extension,
+    SearchEngine,
+    SearchEngineSuggestion,
+    RemoteTab,
+    VisitURL,
+    SearchEngineFavicon,
+    Favicon,
+}
+
+pub struct Match {
```

Comment:
Match, MatchReason, and MatchLabel should probably all #[derive(Debug, Clone)]. Maybe others too. Anything public IMO.